{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TACTIC tutorials\n",
    "\n",
    "This notebook provides quick tutorials on X applications for `TACTIC`: \n",
    "1. Predicting drug interaction outcomes for new strains (using the already built model)\n",
    "2. Predicting drug interaction outcomes for new drug combinations (using the already built model)\n",
    "3. Building a TACTIC model with user-defined inputs\n",
    "\n",
    "# How to use `TACTIC`\n",
    "\n",
    "This notebook provides a quick tutorial on how to apply TACTIC using sample drug response and drug interaction data. The code below is adaptable to other datasets provided that they are formatted according to the [example datasets](). \n",
    "\n",
    "Of note, TACTIC requires `indigopy`, a Python package implementation of [INDIGO](https://doi.org/10.1007/978-1-4939-8891-4_13). Instructions on how to install `indigopy` are detailed in the package [README](https://github.com/sriram-lab/INDIGOpy/blob/main/README.md)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Set up environment\n",
    "\n",
    "Import dependencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data handling\n",
    "import webbrowser\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from itertools import chain, compress\n",
    "from warnings import warn\n",
    "from scipy import sparse\n",
    "from indigopy.core import featurize, classify\n",
    "\n",
    "# Data analysis\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "# Machine learning\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "\n",
    "# Data visualization\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Prepare sample data\n",
    "\n",
    "TACTIC requires 2 types of data inputs: \n",
    "\n",
    "1. Drug response (e.g., omics measurements to drug treatments)\n",
    "2. Drug interactions (e.g., Bliss or Loewe scores)\n",
    "\n",
    "For this tutorial, we will work with subsets of the [data](https://github.com/sriram-lab/TACTIC/tree/master/data) used for developing the TACTIC approach. Specifically: \n",
    "\n",
    "- We will use data from the [Brochado study](https://pubmed.ncbi.nlm.nih.gov/29973719/) for model training, and data measured in *A. baumannii* for model testing\n",
    "- Based on the drugs represented in the sample drug interactions, we will take: \n",
    "    - The subset of chemogenomics data measured for the first 50 genes in *E. coli* MG1655 ([Nichols et al., 2011](https://pubmed.ncbi.nlm.nih.gov/21185072/))\n",
    "    - The subset of transcriptomics data measured for the first 50 genes in *M. tb* H37Rv ([Ma et al., 2019](https://pubmed.ncbi.nlm.nih.gov/31719182/))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample datasets (N = 5):\n",
      "\t1. Drug interactions, train set (192, 9):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>Set</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Interaction</th>\n",
       "      <th>Degree</th>\n",
       "      <th>Score</th>\n",
       "      <th>Label</th>\n",
       "      <th>Metric</th>\n",
       "      <th>Occurrence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>E. coli BW25113</td>\n",
       "      <td>A22, MMC</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.154161</td>\n",
       "      <td>Synergy</td>\n",
       "      <td>Bliss</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>E. coli BW25113</td>\n",
       "      <td>AMK, EGCG</td>\n",
       "      <td>2</td>\n",
       "      <td>0.134583</td>\n",
       "      <td>Antagonism</td>\n",
       "      <td>Bliss</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>E. coli BW25113</td>\n",
       "      <td>AMK, ERY</td>\n",
       "      <td>2</td>\n",
       "      <td>0.200988</td>\n",
       "      <td>Antagonism</td>\n",
       "      <td>Bliss</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>E. coli BW25113</td>\n",
       "      <td>AMK, PRC</td>\n",
       "      <td>2</td>\n",
       "      <td>0.472862</td>\n",
       "      <td>Antagonism</td>\n",
       "      <td>Bliss</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>E. coli BW25113</td>\n",
       "      <td>AMK, AMX</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.128012</td>\n",
       "      <td>Synergy</td>\n",
       "      <td>Bliss</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Source     Set           Strain Interaction  Degree     Score       Label  \\\n",
       "2        1  TACTIC  E. coli BW25113    A22, MMC       2 -0.154161     Synergy   \n",
       "8        1  TACTIC  E. coli BW25113   AMK, EGCG       2  0.134583  Antagonism   \n",
       "9        1  TACTIC  E. coli BW25113    AMK, ERY       2  0.200988  Antagonism   \n",
       "14       1  TACTIC  E. coli BW25113    AMK, PRC       2  0.472862  Antagonism   \n",
       "21       1  TACTIC  E. coli BW25113    AMK, AMX       2 -0.128012     Synergy   \n",
       "\n",
       "   Metric  Occurrence  \n",
       "2   Bliss           6  \n",
       "8   Bliss           6  \n",
       "9   Bliss           9  \n",
       "14  Bliss           6  \n",
       "21  Bliss           8  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t2. Drug interactions, test set (45, 9):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>Set</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Interaction</th>\n",
       "      <th>Degree</th>\n",
       "      <th>Score</th>\n",
       "      <th>Label</th>\n",
       "      <th>Metric</th>\n",
       "      <th>Occurrence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1850</th>\n",
       "      <td>3</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>A. baumannii ATCC 17978</td>\n",
       "      <td>AMK, AMP</td>\n",
       "      <td>2</td>\n",
       "      <td>0.317064</td>\n",
       "      <td>Antagonism</td>\n",
       "      <td>Loewe</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1851</th>\n",
       "      <td>3</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>A. baumannii ATCC 17978</td>\n",
       "      <td>AMK, AZT</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.606979</td>\n",
       "      <td>Synergy</td>\n",
       "      <td>Loewe</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1852</th>\n",
       "      <td>3</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>A. baumannii ATCC 17978</td>\n",
       "      <td>AMK, NAL</td>\n",
       "      <td>2</td>\n",
       "      <td>0.246822</td>\n",
       "      <td>Antagonism</td>\n",
       "      <td>Loewe</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>3</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>A. baumannii ATCC 17978</td>\n",
       "      <td>AMK, RIF</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.132424</td>\n",
       "      <td>Neutral</td>\n",
       "      <td>Loewe</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>3</td>\n",
       "      <td>TACTIC</td>\n",
       "      <td>A. baumannii ATCC 17978</td>\n",
       "      <td>AMK, TET</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.458472</td>\n",
       "      <td>Synergy</td>\n",
       "      <td>Loewe</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Source     Set                   Strain Interaction  Degree     Score  \\\n",
       "1850       3  TACTIC  A. baumannii ATCC 17978    AMK, AMP       2  0.317064   \n",
       "1851       3  TACTIC  A. baumannii ATCC 17978    AMK, AZT       2 -0.606979   \n",
       "1852       3  TACTIC  A. baumannii ATCC 17978    AMK, NAL       2  0.246822   \n",
       "1853       3  TACTIC  A. baumannii ATCC 17978    AMK, RIF       2 -0.132424   \n",
       "1854       3  TACTIC  A. baumannii ATCC 17978    AMK, TET       2 -0.458472   \n",
       "\n",
       "           Label Metric  Occurrence  \n",
       "1850  Antagonism  Loewe           3  \n",
       "1851     Synergy  Loewe           5  \n",
       "1852  Antagonism  Loewe           4  \n",
       "1853     Neutral  Loewe           6  \n",
       "1854     Synergy  Loewe           4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t3. Drug response, key (31, 3):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Code</th>\n",
       "      <th>MG1655</th>\n",
       "      <th>H37Rv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A22</td>\n",
       "      <td>A22-15.0</td>\n",
       "      <td>A22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMK</td>\n",
       "      <td>AMIKACIN-0.2</td>\n",
       "      <td>AMIKACIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMX</td>\n",
       "      <td>AMOXICILLIN-1.5</td>\n",
       "      <td>AMOXICILLIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMP</td>\n",
       "      <td>AMPICILLIN-8.0</td>\n",
       "      <td>AMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AZI</td>\n",
       "      <td>AZITHROMYCIN-1.0</td>\n",
       "      <td>AZITHROMYCIN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Code            MG1655         H37Rv\n",
       "0  A22          A22-15.0           A22\n",
       "1  AMK      AMIKACIN-0.2      AMIKACIN\n",
       "2  AMX   AMOXICILLIN-1.5   AMOXICILLIN\n",
       "3  AMP    AMPICILLIN-8.0           AMP\n",
       "4  AZI  AZITHROMYCIN-1.0  AZITHROMYCIN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t4. Drug response, MG1655 (50, 32):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>A22-15.0</th>\n",
       "      <th>AMIKACIN-0.2</th>\n",
       "      <th>AMOXICILLIN-1.5</th>\n",
       "      <th>AMPICILLIN-8.0</th>\n",
       "      <th>AZITHROMYCIN-1.0</th>\n",
       "      <th>AZTREONAM-0.04</th>\n",
       "      <th>BACITRACIN-300</th>\n",
       "      <th>BENZALKONIUM-25</th>\n",
       "      <th>CCCP-2.0</th>\n",
       "      <th>...</th>\n",
       "      <th>MITOMYCINC-0.1</th>\n",
       "      <th>NALIDIXICACID-2.0</th>\n",
       "      <th>NOVOBIOCIN-30</th>\n",
       "      <th>PARAQUAT-18.0</th>\n",
       "      <th>POLYMYXINB-6.0</th>\n",
       "      <th>PROCAINE-30</th>\n",
       "      <th>RIFAMPICIN-2.0</th>\n",
       "      <th>SPECTINOMYCIN-6.0</th>\n",
       "      <th>TETRACYCLINE-1.0</th>\n",
       "      <th>VERAPAMIL-1.0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b0002</td>\n",
       "      <td>-0.559545</td>\n",
       "      <td>1.647844</td>\n",
       "      <td>-0.994845</td>\n",
       "      <td>-1.298203</td>\n",
       "      <td>-0.468216</td>\n",
       "      <td>-1.332281</td>\n",
       "      <td>1.796933</td>\n",
       "      <td>1.178283</td>\n",
       "      <td>-1.030014</td>\n",
       "      <td>...</td>\n",
       "      <td>1.945560</td>\n",
       "      <td>1.462836</td>\n",
       "      <td>1.138723</td>\n",
       "      <td>-4.531073</td>\n",
       "      <td>2.045477</td>\n",
       "      <td>0.296932</td>\n",
       "      <td>-0.265948</td>\n",
       "      <td>-1.751131</td>\n",
       "      <td>-0.507021</td>\n",
       "      <td>0.395577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b0003</td>\n",
       "      <td>0.242666</td>\n",
       "      <td>-0.693398</td>\n",
       "      <td>-1.234871</td>\n",
       "      <td>0.411192</td>\n",
       "      <td>0.285870</td>\n",
       "      <td>0.469039</td>\n",
       "      <td>1.043688</td>\n",
       "      <td>0.669880</td>\n",
       "      <td>-0.567867</td>\n",
       "      <td>...</td>\n",
       "      <td>0.620079</td>\n",
       "      <td>1.579996</td>\n",
       "      <td>0.008923</td>\n",
       "      <td>-1.120873</td>\n",
       "      <td>2.009723</td>\n",
       "      <td>0.325243</td>\n",
       "      <td>-0.435470</td>\n",
       "      <td>0.357894</td>\n",
       "      <td>0.483735</td>\n",
       "      <td>0.599244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b0004</td>\n",
       "      <td>1.809979</td>\n",
       "      <td>1.482407</td>\n",
       "      <td>0.687448</td>\n",
       "      <td>0.695141</td>\n",
       "      <td>0.367449</td>\n",
       "      <td>0.135136</td>\n",
       "      <td>-0.680992</td>\n",
       "      <td>1.641956</td>\n",
       "      <td>-0.098073</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.623568</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>-1.053464</td>\n",
       "      <td>0.111009</td>\n",
       "      <td>1.346476</td>\n",
       "      <td>-1.616272</td>\n",
       "      <td>0.805449</td>\n",
       "      <td>-1.930057</td>\n",
       "      <td>0.384499</td>\n",
       "      <td>1.035534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0005</td>\n",
       "      <td>-0.625091</td>\n",
       "      <td>0.140624</td>\n",
       "      <td>1.369324</td>\n",
       "      <td>0.780893</td>\n",
       "      <td>0.084454</td>\n",
       "      <td>0.090245</td>\n",
       "      <td>-0.826658</td>\n",
       "      <td>1.112208</td>\n",
       "      <td>0.650183</td>\n",
       "      <td>...</td>\n",
       "      <td>0.633023</td>\n",
       "      <td>0.099400</td>\n",
       "      <td>1.897863</td>\n",
       "      <td>0.318214</td>\n",
       "      <td>-0.733713</td>\n",
       "      <td>-1.782458</td>\n",
       "      <td>-0.149242</td>\n",
       "      <td>-0.814743</td>\n",
       "      <td>0.353229</td>\n",
       "      <td>-1.520680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b0006</td>\n",
       "      <td>-0.357582</td>\n",
       "      <td>-0.134298</td>\n",
       "      <td>-0.649026</td>\n",
       "      <td>1.327997</td>\n",
       "      <td>-0.260863</td>\n",
       "      <td>-1.039351</td>\n",
       "      <td>0.019330</td>\n",
       "      <td>-0.458028</td>\n",
       "      <td>0.238005</td>\n",
       "      <td>...</td>\n",
       "      <td>0.467696</td>\n",
       "      <td>0.016489</td>\n",
       "      <td>0.008245</td>\n",
       "      <td>-0.646982</td>\n",
       "      <td>1.304580</td>\n",
       "      <td>0.833593</td>\n",
       "      <td>0.459269</td>\n",
       "      <td>-0.074700</td>\n",
       "      <td>-1.631192</td>\n",
       "      <td>-1.015940</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene  A22-15.0  AMIKACIN-0.2  AMOXICILLIN-1.5  AMPICILLIN-8.0  \\\n",
       "0  b0002 -0.559545      1.647844        -0.994845       -1.298203   \n",
       "1  b0003  0.242666     -0.693398        -1.234871        0.411192   \n",
       "2  b0004  1.809979      1.482407         0.687448        0.695141   \n",
       "3  b0005 -0.625091      0.140624         1.369324        0.780893   \n",
       "4  b0006 -0.357582     -0.134298        -0.649026        1.327997   \n",
       "\n",
       "   AZITHROMYCIN-1.0  AZTREONAM-0.04  BACITRACIN-300  BENZALKONIUM-25  \\\n",
       "0         -0.468216       -1.332281        1.796933         1.178283   \n",
       "1          0.285870        0.469039        1.043688         0.669880   \n",
       "2          0.367449        0.135136       -0.680992         1.641956   \n",
       "3          0.084454        0.090245       -0.826658         1.112208   \n",
       "4         -0.260863       -1.039351        0.019330        -0.458028   \n",
       "\n",
       "   CCCP-2.0  ...  MITOMYCINC-0.1  NALIDIXICACID-2.0  NOVOBIOCIN-30  \\\n",
       "0 -1.030014  ...        1.945560           1.462836       1.138723   \n",
       "1 -0.567867  ...        0.620079           1.579996       0.008923   \n",
       "2 -0.098073  ...       -1.623568           0.000469      -1.053464   \n",
       "3  0.650183  ...        0.633023           0.099400       1.897863   \n",
       "4  0.238005  ...        0.467696           0.016489       0.008245   \n",
       "\n",
       "   PARAQUAT-18.0  POLYMYXINB-6.0  PROCAINE-30  RIFAMPICIN-2.0  \\\n",
       "0      -4.531073        2.045477     0.296932       -0.265948   \n",
       "1      -1.120873        2.009723     0.325243       -0.435470   \n",
       "2       0.111009        1.346476    -1.616272        0.805449   \n",
       "3       0.318214       -0.733713    -1.782458       -0.149242   \n",
       "4      -0.646982        1.304580     0.833593        0.459269   \n",
       "\n",
       "   SPECTINOMYCIN-6.0  TETRACYCLINE-1.0  VERAPAMIL-1.0  \n",
       "0          -1.751131         -0.507021       0.395577  \n",
       "1           0.357894          0.483735       0.599244  \n",
       "2          -1.930057          0.384499       1.035534  \n",
       "3          -0.814743          0.353229      -1.520680  \n",
       "4          -0.074700         -1.631192      -1.015940  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t5. Drug response, H37Rv(50, 32):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>A22</th>\n",
       "      <th>AMIKACIN</th>\n",
       "      <th>AMOXICILLIN</th>\n",
       "      <th>AMP</th>\n",
       "      <th>AZITHROMYCIN</th>\n",
       "      <th>AZTREONAM</th>\n",
       "      <th>BACITRACIN</th>\n",
       "      <th>BENZALKONIUM</th>\n",
       "      <th>CCCP</th>\n",
       "      <th>...</th>\n",
       "      <th>MTM</th>\n",
       "      <th>NALIDIXICACID</th>\n",
       "      <th>NOVOBIOCIN</th>\n",
       "      <th>PARAQUAT</th>\n",
       "      <th>POLYMYXINB</th>\n",
       "      <th>PROCAINE</th>\n",
       "      <th>RIF</th>\n",
       "      <th>SPECTINOMYCIN</th>\n",
       "      <th>TET</th>\n",
       "      <th>VERAPAMIL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rv0001</td>\n",
       "      <td>0.280216</td>\n",
       "      <td>-0.435491</td>\n",
       "      <td>0.874587</td>\n",
       "      <td>0.769623</td>\n",
       "      <td>-0.597280</td>\n",
       "      <td>-0.409214</td>\n",
       "      <td>0.455288</td>\n",
       "      <td>0.039375</td>\n",
       "      <td>-1.716496</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012251</td>\n",
       "      <td>-1.218246</td>\n",
       "      <td>-0.502595</td>\n",
       "      <td>-0.374935</td>\n",
       "      <td>0.247419</td>\n",
       "      <td>-0.093330</td>\n",
       "      <td>-0.684493</td>\n",
       "      <td>0.793279</td>\n",
       "      <td>-0.839782</td>\n",
       "      <td>1.710830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rv0002</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>0.224606</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.562953</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.422903</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007786</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.667940</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>-0.700379</td>\n",
       "      <td>-0.294181</td>\n",
       "      <td>0.502027</td>\n",
       "      <td>-0.172280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rv0003</td>\n",
       "      <td>-0.478872</td>\n",
       "      <td>0.300680</td>\n",
       "      <td>0.173256</td>\n",
       "      <td>-0.654577</td>\n",
       "      <td>-1.029745</td>\n",
       "      <td>0.120626</td>\n",
       "      <td>0.107939</td>\n",
       "      <td>-0.587100</td>\n",
       "      <td>0.111387</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.161867</td>\n",
       "      <td>-0.603865</td>\n",
       "      <td>-0.176118</td>\n",
       "      <td>0.407015</td>\n",
       "      <td>-0.138354</td>\n",
       "      <td>0.333373</td>\n",
       "      <td>-0.996568</td>\n",
       "      <td>-0.739771</td>\n",
       "      <td>1.053258</td>\n",
       "      <td>-1.171968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rv0004</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.487582</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>-0.046152</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>-0.172340</td>\n",
       "      <td>...</td>\n",
       "      <td>0.133206</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.089403</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.053454</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>-0.568754</td>\n",
       "      <td>0.129359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rv0005</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>-1.297794</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>0.433942</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>-0.002451</td>\n",
       "      <td>...</td>\n",
       "      <td>0.173396</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>-0.396215</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>-0.317155</td>\n",
       "      <td>0.137907</td>\n",
       "      <td>-1.039578</td>\n",
       "      <td>-0.061960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene       A22  AMIKACIN  AMOXICILLIN       AMP  AZITHROMYCIN  AZTREONAM  \\\n",
       "0  Rv0001  0.280216 -0.435491     0.874587  0.769623     -0.597280  -0.409214   \n",
       "1  Rv0002 -0.294181  0.224606    -0.294181 -0.562953     -0.294181  -0.294181   \n",
       "2  Rv0003 -0.478872  0.300680     0.173256 -0.654577     -1.029745   0.120626   \n",
       "3  Rv0004  0.004665  0.487582     0.004665 -0.046152      0.004665   0.004665   \n",
       "4  Rv0005  0.137907 -1.297794     0.137907  0.433942      0.137907   0.137907   \n",
       "\n",
       "   BACITRACIN  BENZALKONIUM      CCCP  ...       MTM  NALIDIXICACID  \\\n",
       "0    0.455288      0.039375 -1.716496  ...  0.012251      -1.218246   \n",
       "1   -0.294181     -0.294181 -0.422903  ...  0.007786      -0.294181   \n",
       "2    0.107939     -0.587100  0.111387  ... -0.161867      -0.603865   \n",
       "3    0.004665      0.004665 -0.172340  ...  0.133206       0.004665   \n",
       "4    0.137907      0.137907 -0.002451  ...  0.173396       0.137907   \n",
       "\n",
       "   NOVOBIOCIN  PARAQUAT  POLYMYXINB  PROCAINE       RIF  SPECTINOMYCIN  \\\n",
       "0   -0.502595 -0.374935    0.247419 -0.093330 -0.684493       0.793279   \n",
       "1   -0.667940 -0.294181   -0.294181 -0.294181 -0.700379      -0.294181   \n",
       "2   -0.176118  0.407015   -0.138354  0.333373 -0.996568      -0.739771   \n",
       "3    0.089403  0.004665    0.004665  0.004665  0.053454       0.004665   \n",
       "4   -0.396215  0.137907    0.137907  0.137907 -0.317155       0.137907   \n",
       "\n",
       "        TET  VERAPAMIL  \n",
       "0 -0.839782   1.710830  \n",
       "1  0.502027  -0.172280  \n",
       "2  1.053258  -1.171968  \n",
       "3 -0.568754   0.129359  \n",
       "4 -1.039578  -0.061960  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully saved drug response data for key\n",
      "Successfully saved drug response data for MG1655\n",
      "Successfully saved drug response data for H37Rv\n",
      "Successfully saved drug interaction data for train\n",
      "Successfully saved drug interaction data for test\n"
     ]
    }
   ],
   "source": [
    "# Instantiate data variable\n",
    "data = {'source': {}, 'response': {}, 'interactions': {}}\n",
    "\n",
    "# Load source data\n",
    "data['source']['interactions']  = pd.read_excel('./data/ixn_data.xlsx', sheet_name='data', engine='openpyxl')\n",
    "data['source']['key']           = pd.read_excel('./data/ixn_data.xlsx', sheet_name='drug_key', engine='openpyxl')\n",
    "data['source']['MG1655']        = pd.read_excel('./data/omics_data.xlsx', sheet_name='ecoli_chemogenomics', engine='openpyxl')\n",
    "data['source']['H37Rv']         = pd.read_excel('./data/omics_data.xlsx', sheet_name='mtb_transcriptomics', engine='openpyxl')\n",
    "\n",
    "# Prepare data: drug interactions (train and test sets)\n",
    "dfi = data['source']['interactions'].copy()\n",
    "dfo = dfi[dfi['Source']==1]\n",
    "ix = dfo['Interaction'].value_counts()==6\n",
    "ixns = list(ix.index[ix==True])\n",
    "data['interactions']['train']   = dfo[dfo['Interaction'].isin(ixns)]\n",
    "data['interactions']['test']    = dfi[dfi['Strain'].str.startswith('A.')]\n",
    "\n",
    "# Prepare data: drug response (key, MG1655, H37Rv)\n",
    "dfi, df1, df2 = data['source']['key'].copy(), data['source']['MG1655'].copy(), data['source']['H37Rv'].copy()\n",
    "ixns = ixns + data['interactions']['test']['Interaction'].tolist()\n",
    "drugs = sorted(list(set(chain.from_iterable([list(combo.split(', ')) for combo in ixns]))))\n",
    "dfo = dfi[dfi['Code'].isin(drugs)].rename(columns={'Chemogenomic_label': 'MG1655', 'Transcriptomic_label': 'H37Rv'})\n",
    "data['response']['key']         = dfo.copy()\n",
    "data['response']['MG1655']      = df1[['Gene'] + dfo['MG1655'].tolist()].iloc[:50, :]\n",
    "data['response']['H37Rv']       = df2[['Gene'] + dfo['H37Rv'].tolist()].iloc[:50, :]\n",
    "\n",
    "# Inspect sample data\n",
    "print('Sample datasets (N = 5):')\n",
    "df = data['interactions']['train'].copy()\n",
    "print('\\t1. Drug interactions, train set {}:'.format(df.shape))\n",
    "display(df.head())\n",
    "df = data['interactions']['test'].copy()\n",
    "print('\\t2. Drug interactions, test set {}:'.format(df.shape))\n",
    "display(df.head())\n",
    "df = data['response']['key'].copy()\n",
    "print('\\t3. Drug response, key {}:'.format(df.shape))\n",
    "display(df.head())\n",
    "df = data['response']['MG1655'].copy()\n",
    "print('\\t4. Drug response, MG1655 {}:'.format(df.shape))\n",
    "display(df.head())\n",
    "df = data['response']['H37Rv'].copy()\n",
    "print('\\t5. Drug response, H37Rv{}:'.format(df.shape))\n",
    "display(df.head())\n",
    "\n",
    "# Save data for external inspection\n",
    "fname = './data/sample/drug_response.xlsx'\n",
    "for key, value in data['response'].items(): \n",
    "\tif key=='key': \n",
    "\t\tvalue.to_excel(fname, sheet_name=key, index=False, engine='openpyxl', freeze_panes=(1, 1))\n",
    "\telse: \n",
    "\t\twith pd.ExcelWriter(fname, mode='a', if_sheet_exists='replace') as file:\n",
    "\t\t\tvalue.to_excel(file, sheet_name=key, index=False, engine='openpyxl', freeze_panes=(1, 1))\n",
    "\tprint('Successfully saved drug response data for {}'.format(key))\n",
    "fname = './data/sample/drug_interactions.xlsx'\n",
    "for key, value in data['interactions'].items(): \n",
    "\tif key=='train': \n",
    "\t\tvalue.to_excel(fname, sheet_name=key, index=False, engine='openpyxl', freeze_panes=(1, 1))\n",
    "\telse: \n",
    "\t\twith pd.ExcelWriter(fname, mode='a', if_sheet_exists='replace') as file:\n",
    "\t\t\tvalue.to_excel(file, sheet_name=key, index=False, engine='openpyxl', freeze_panes=(1, 1))\n",
    "\tprint('Successfully saved drug interaction data for {}'.format(key))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Obtain gene orthologs via OrtholugeDB\n",
    "\n",
    "One of the key concepts integrated in TACTIC is **transfer learning via orthology mapping**, which extends drug response data measured in a reference organism (e.g., *E. coli* MG1655) to a new organism (e.g., *A. baumannii*). To do so, information on gene orthologs shared between the reference and new organisms must be available. In this section, we walk through how to extract gene orthology data from [OrtholugeDB](https://ortholugedb.ca/). The code below opens the [webpage](https://ortholugedb.ca/?page=matrix-setup) where we can specify the reference and comparison genomes: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Open genome comparison webpage in OrtholugeDB\n",
    "webbrowser.open('https://ortholugedb.ca/?page=matrix-setup')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div>\n",
    "<img src='./data/sample/images/0_OrtholugeDB_setup.png' width='1000'/>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this tutorial, our reference genomes are *E. coli* MG1655 and *M. tb* H37Rv. We will use the webpage above to collect gene orthology information for both references. Starting with *E. coli* MG1655, we can type the keyword `mg1655` to quickly find the reference genome: \n",
    "\n",
    "<div>\n",
    "<img src='./data/sample/images/1_MG1655_reference.png' width='1000'/>\n",
    "</div>\n",
    "\n",
    "Based on our sample dataset, there are 9 strains that we are interested in: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Strains of interest (N = 9):\n",
      "\tE. coli MG1655\n",
      "\tM. tb H37Rv\n",
      "\tE. coli BW25113\n",
      "\tE. coli iAi1\n",
      "\tP. aeruginosa PAO1\n",
      "\tP. aeruginosa PA14\n",
      "\tS. typhimurium 14028s\n",
      "\tS. typhimurium LT2\n",
      "\tA. baumannii ATCC 17978\n"
     ]
    }
   ],
   "source": [
    "# Print out strains of interest\n",
    "df = pd.concat([data['interactions']['train'], data['interactions']['test']])\n",
    "strains = ['E. coli MG1655', 'M. tb H37Rv'] + list(df['Strain'].unique())\n",
    "print('Strains of interest (N = {}):\\n\\t{}'.format(len(strains), '\\n\\t'.join(strains)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Against *E. coli* MG1655 as the reference genome, we can find genomes for the remaining 8 strains using the following keywords: \n",
    "\n",
    "- *M. tb* H37Rv: `h37rv`\n",
    "- *E. coli* iAi1: `iai1`\n",
    "- *P. aeruginosa* PAO1: `pao1`\n",
    "- *P. aeruginosa* PA14: `pa14`\n",
    "- *S.* Typhimurium 14028s: `14028s`\n",
    "- *S.* Typhimurium LT2: `lt2`\n",
    "- *A. baumannii* ATCC 17978: `17978`\n",
    "\n",
    "Of note, the genome for *E. coli* BW25113 is not available through OrtholugeDB. Given that both BW25113 and MG1655 are [derived from *E. coli* K-12](https://ecoliwiki.org/colipedia/index.php/Category:Strain:BW25113), we assume that their genomes are equivalent.\n",
    "\n",
    "<div>\n",
    "<img src='./data/sample/images/2_H37Rv_comparison.png' width='1000'/>\n",
    "</div>\n",
    "\n",
    "<div>\n",
    "<img src='./data/sample/images/3_all_comparisons.png' width='1000'/>\n",
    "</div>\n",
    "\n",
    "Once we have all genomes of interest selected, we can advance to the next step where we specify the level of orthology information desired. For TACTIC, we recommend leaving the default options: \n",
    "\n",
    "<div>\n",
    "<img src='./data/sample/images/4_gene_filter.png' width='1000'/>\n",
    "</div>\n",
    "\n",
    "The results page should look like the following: \n",
    "\n",
    "<div>\n",
    "<img src='./data/sample/images/5_results_page.png' width='1000'/>\n",
    "</div>\n",
    "\n",
    "We can download the gene orthology data for *E. coli* MG1655 in either CSV or TAB (recommended) format. To collect gene orthology information for *M. tb* H37Rv, we can go back to step 1 and use `h37rv` as our reference genome keyword: \n",
    "\n",
    "<div>\n",
    "<img src='./data/sample/images/6_H37Rv_reference.png' width='1000'/>\n",
    "</div>\n",
    "\n",
    "We also select *E. coli* MG1655 as one of our comparison genomes (keyword: `mg1655`) and de-select *M. tb* H37Rv: \n",
    "\n",
    "<div>\n",
    "<img src='./data/sample/images/7_MG1655_comparison.png' width='1000'/>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Create orthology maps\n",
    "\n",
    "Using the downloaded files from OrtholugeDB, we can define orhtology maps for each reference strain in the following manner: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Queried strain (E. coli MG1655) is the same as reference (MG1655). Continuing...\n",
      "Number of orthologs for M. tb H37Rv against MG1655 reference: 1124\n",
      "Number of orthologs for E. coli BW25113 against MG1655 reference: 4157\n",
      "Number of orthologs for E. coli iAi1 against MG1655 reference: 3772\n",
      "Number of orthologs for P. aeruginosa PAO1 against MG1655 reference: 2078\n",
      "Number of orthologs for P. aeruginosa PA14 against MG1655 reference: 2083\n",
      "Number of orthologs for S. typhimurium 14028s against MG1655 reference: 3206\n",
      "Number of orthologs for S. typhimurium LT2 against MG1655 reference: 3193\n",
      "Number of orthologs for A. baumannii ATCC 17978 against MG1655 reference: 1563\n",
      "Number of orthologs for E. coli MG1655 against H37Rv reference: 1064\n",
      "Queried strain (M. tb H37Rv) is the same as reference (H37Rv). Continuing...\n",
      "Number of orthologs for E. coli BW25113 against H37Rv reference: 1064\n",
      "Number of orthologs for E. coli iAi1 against H37Rv reference: 1056\n",
      "Number of orthologs for P. aeruginosa PAO1 against H37Rv reference: 1223\n",
      "Number of orthologs for P. aeruginosa PA14 against H37Rv reference: 1231\n",
      "Number of orthologs for S. typhimurium 14028s against H37Rv reference: 1044\n",
      "Number of orthologs for S. typhimurium LT2 against H37Rv reference: 1036\n",
      "Number of orthologs for A. baumannii ATCC 17978 against H37Rv reference: 1020\n"
     ]
    }
   ],
   "source": [
    "# Load orthology data\n",
    "data['source']['orthology'] = {\n",
    "    'MG1655': pd.read_csv('./data/sample/MG1655_orthologs.txt', sep='\\t'), \n",
    "    'H37Rv': pd.read_csv('./data/sample/H37Rv_orthologs.txt', sep='\\t')\n",
    "}\n",
    "\n",
    "# Define orthology map for each reference\n",
    "data['orthology'] = {}\n",
    "for key, value in data['source']['orthology'].items(): \n",
    "    # Re-format orthology data\n",
    "    jx = [1, 3] + [x for x in range(4, value.shape[1], 3)]\n",
    "    df = value.copy().iloc[:, jx]\n",
    "    df.columns = [s.split(' - ')[0].upper() if 'comparison' in s else s for s in df.columns]\n",
    "    # Extra step for M. tb H37Rv\n",
    "    df.iloc[:, 0] = [s.replace('RVBD_', 'Rv') for s in df.iloc[:, 0]]\n",
    "    # Create orthology map\n",
    "    data['orthology'][key] = {}\n",
    "    for strain in strains: \n",
    "        # Define keyword\n",
    "        keyword = strain.split(' ')[-1].upper()\n",
    "        # Account for special cases\n",
    "        if key.upper()==keyword: \n",
    "            print('Queried strain ({}) is the same as reference ({}). Continuing...'.format(strain, key))\n",
    "            continue\n",
    "        elif (keyword=='BW25113') and (key=='MG1655'): \n",
    "            data['orthology'][key][strain] = df.iloc[:, 0].tolist()\n",
    "        else: \n",
    "            if keyword=='BW25113': \n",
    "                keyword = 'MG1655'\n",
    "            jx = df.columns.str.endswith(keyword)\n",
    "            if not any(jx): \n",
    "                warn('No orthologs found for {} against {} reference'.format(strain, key))\n",
    "                continue\n",
    "            else: \n",
    "                data['orthology'][key][strain] = list(compress(df.iloc[:, 0], (df.iloc[:, jx] != '0').values))\n",
    "        # Print total number of orthologs\n",
    "        print('Number of orthologs for {} against {} reference: {}'.format(strain, key, len(data['orthology'][key][strain])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Data featurization using `indigopy`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. ML model training and testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
